File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.java
class Solution {
    private static final int MOD = (int) 1e9 + 7;

    public int numberOfSets(int n, int k) {
        // dp[i][j] = number of ways to form j segments using first i points
        // where point i is NOT the end of a segment
        int[][] dpNotEndingAtPoint = new int[n + 1][k + 1];
      
        // dpEndingAtPoint[i][j] = number of ways to form j segments using first i points
        // where point i IS the end of a segment
        int[][] dpEndingAtPoint = new int[n + 1][k + 1];
      
        // Base case: with 1 point and 0 segments, there's exactly 1 way
        dpNotEndingAtPoint[1][0] = 1;
      
        // Fill the DP tables for each point position
        for (int currentPoint = 2; currentPoint <= n; ++currentPoint) {
            for (int segmentCount = 0; segmentCount <= k; ++segmentCount) {
                // Case 1: Current point is not the end of a segment
                // We can transition from previous point being either end or not end
                dpNotEndingAtPoint[currentPoint][segmentCount] = 
                    (dpNotEndingAtPoint[currentPoint - 1][segmentCount] + 
                     dpEndingAtPoint[currentPoint - 1][segmentCount]) % MOD;
              
                // Case 2: Current point is the end of a segment
                // First, extend existing segments that end at previous point
                dpEndingAtPoint[currentPoint][segmentCount] = 
                    dpEndingAtPoint[currentPoint - 1][segmentCount];
              
                // If we have segments to form, we can create a new segment ending here
                if (segmentCount > 0) {
                    // Start a new segment from previous point (not ending)
                    dpEndingAtPoint[currentPoint][segmentCount] = 
                        (dpEndingAtPoint[currentPoint][segmentCount] + 
                         dpNotEndingAtPoint[currentPoint - 1][segmentCount - 1]) % MOD;
                  
                    // Extend a segment that was ending at previous point
                    dpEndingAtPoint[currentPoint][segmentCount] = 
                        (dpEndingAtPoint[currentPoint][segmentCount] + 
                         dpEndingAtPoint[currentPoint - 1][segmentCount - 1]) % MOD;
                }
            }
        }
      
        // Return total ways: sum of both cases at point n with k segments
        return (dpNotEndingAtPoint[n][k] + dpEndingAtPoint[n][k]) % MOD;
    }
}

File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.ts
/**
 * Counts the number of ways to draw k non-overlapping line segments on n points
 * @param n - The number of points on a line
 * @param k - The number of line segments to draw
 * @returns The number of ways modulo 10^9 + 7
 */
function numberOfSets(n: number, k: number): number {
    // dp[i][j] represents the number of ways to draw j segments using first i points
    // where the last segment does NOT end at point i
    const dpNotEnding: number[][] = Array.from(
        { length: n + 1 }, 
        () => new Array(k + 1).fill(0)
    );
  
    // dp[i][j] represents the number of ways to draw j segments using first i points
    // where the last segment DOES end at point i
    const dpEnding: number[][] = Array.from(
        { length: n + 1 }, 
        () => new Array(k + 1).fill(0)
    );
  
    // Base case: with 1 point and 0 segments, there's 1 way (empty set)
    dpNotEnding[1][0] = 1;
  
    const MOD: number = 10 ** 9 + 7;
  
    // Iterate through each point position
    for (let i = 2; i <= n; ++i) {
        // Iterate through each possible number of segments
        for (let j = 0; j <= k; ++j) {
            // Ways where j segments don't end at point i = 
            // ways from previous point (either ending or not ending there)
            dpNotEnding[i][j] = (dpNotEnding[i - 1][j] + dpEnding[i - 1][j]) % MOD;
          
            // Extend segments that ended at point i-1 to point i
            dpEnding[i][j] = dpEnding[i - 1][j];
          
            if (j > 0) {
                // Start a new segment at point i (from any previous configuration with j-1 segments)
                dpEnding[i][j] = (dpEnding[i][j] + dpNotEnding[i - 1][j - 1]) % MOD;
              
                // Extend a segment that was already ending at i-1 with j-1 segments
                dpEnding[i][j] = (dpEnding[i][j] + dpEnding[i - 1][j - 1]) % MOD;
            }
        }
    }
  
    // Return total ways: segments either ending or not ending at point n
    return (dpNotEnding[n][k] + dpEnding[n][k]) % MOD;
}
